hadoop-hdds/docs/content/design/leader-execution/leader-execution.md

```diff
@@ -0,0 +1,422 @@
+  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License. See accompanying LICENSE file.
+-->
+
+# Background
+
+Here is the summary of the challenges:
+
+- The current implementation depends on consensus on the order of requests received and not on consensus on the processing of the requests.
+- The double buffer implementation currently is meant to optimize the rate at which writes get flushed to RocksDB but the effective batching achieved is 1.2 at best. It is also a source of continuous bugs and added complexity for new features.
+- The number of transactions that can be pushed through Ratis currently caps out around 25k.
+- The Current performance envelope for OM is around 12k transactions per second. The early testing pushes this to 40k transactions per second.
+
+## Execution at leader node needs deal with below cases
+1. Parallel execution: ratis serialize all the execution in order. With control, it is possible to execute the request in parallel which are independent.
+2. Optimized locking: Locks are taken at bucket level for both read and write flow. Here, focus to remove lock between read and write flow, and have more granular locking.
+3. Cache Optimization: Cache are maintained for write operation and read also make use of same for consistency. This creates complexity for read to provide accurate result with parallel operation.
+4. Double buffer code complexity: Double buffer provides batching for db update. This is done with ratis state machine and induces issues managing ratis state machine, cache and db updates.
+5. Request execution flow optimization: Optimize request execution flow, removing un-necessary operation and improve testability.
+6. Performance and resource Optimization: Currently, same execution is repeated at all nodes, and have more failure points. With leader side execution and parallelism, need improve performance and resource utilization.
+
+### Object ID generation
+Currently, the Object ID is tied to Ratis transaction metadata. This has multiple challenges in the long run.
+
+- If OM adopts multi Ratis to scale writes further, Object IDs will not longer be unique.
+- If we shard OM, then across OMs the object ID will not be unique.
+- When batching multiple requests, we cannot utilize Ratis metadata to generate object IDs.
+
+Longer term, we should move to a UUID based object ID generation. This will allow us to generate object IDs that are globally unique. In the mean time, we are moving to a persistent counter based object ID generation. The counter is persisted during apply transaction and is incremented for each new object created.
```

Comment:
How about update ID? In the future if we decide to shard the OMs, we probably need some kind of sequence generator to generate a monotonically increasing ID for each update.

Reply:
updateId is just represent change in Existing object or like versioning. So this may be just,` lastUpdateId + 1`

This part is not included as part of this feature and can be done with approach like above or any distributed indexing if required.